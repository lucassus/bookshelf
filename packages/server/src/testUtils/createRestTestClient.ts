import express from "express";
import request from "supertest";

import { User } from "../database/entity/User";
import { routes } from "../rest";

export function createRestTestClient({
  currentUser
}: { currentUser?: User } = {}) {
  const app = express();

  if (currentUser) {
    app.use((req, res, next) => {
      (req as any).user = currentUser;
      next();
    });
  }

  app.use("/", routes);

  return request(app);
}
