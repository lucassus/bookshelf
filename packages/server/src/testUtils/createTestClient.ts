import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";

import { buildContext, Context } from "../graphql/context";
import { rootSchema } from "../graphql/rootSchema";

export function createTestClient(
  contextExtra: Partial<Context> = {}
): ApolloServerTestClient {
  const server = new ApolloServer({
    schema: rootSchema,
    context: buildContext(contextExtra)
  });

  return createApolloTestClient(server);
}
