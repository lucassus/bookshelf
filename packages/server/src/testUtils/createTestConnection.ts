import path from "path";
import { createConnection } from "typeorm";

export const createTestConnection = () =>
  createConnection({
    type: "sqlite",
    database: ":memory:",
    dropSchema: true,
    entities: [path.join(__dirname, "../entity/**/*.ts")],
    synchronize: true,
    logging: false
  });
