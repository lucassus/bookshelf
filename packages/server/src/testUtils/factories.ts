import faker from "faker";
import { getManager, ObjectType, DeepPartial } from "typeorm";

import { hashPassword } from "../common/authentication";
import { titleizeSentence } from "../common/strings";
import { Author } from "../database/entity/Author";
import { Avatar, AVATAR_COLORS } from "../database/entity/Avatar";
import { Book } from "../database/entity/Book";
import { BookCopy } from "../database/entity/BookCopy";
import { User } from "../database/entity/User";

const AVATAR_IMAGES = [
  "/images/avatars/w13.png",
  "/images/avatars/m10.png",
  "/images/avatars/w2.png",
  "/images/avatars/m25.png"
];

const AUTHOR_PHOTOS = [
  "/images/book-authors/j-k-rowling.jpg",
  "/images/book-authors/james-s-a-corey.jpg",
  "/images/book-authors/andrzej-sapkowski.jpg"
];

const BOOK_COVERS = [
  "/images/book-covers/harry1.jpg",
  "/images/book-covers/harry2.jpg",
  "/images/book-covers/harry3.jpg",
  "/images/book-covers/harry4.jpg",
  "/images/book-covers/harry5.jpg",
  "/images/book-covers/harry6.jpg",
  "/images/book-covers/harry7.jpg",
  "/images/book-covers/expanse1.jpg",
  "/images/book-covers/expanse2.jpg",
  "/images/book-covers/expanse3.jpg",
  "/images/book-covers/expanse4.jpg",
  "/images/book-covers/expanse5.jpg",
  "/images/book-covers/expanse6.jpg",
  "/images/book-covers/expanse7.jpg",
  "/images/book-covers/expanse8.jpg",
  "/images/book-covers/witcher1.jpg",
  "/images/book-covers/witcher2.jpg",
  "/images/book-covers/witcher3.jpg",
  "/images/book-covers/witcher4.jpg",
  "/images/book-covers/witcher5.jpg"
];

function createEntity<Entity>(
  entityClass: ObjectType<Entity>,
  attributes: DeepPartial<Entity>
) {
  const manager = getManager();

  const entity = manager.create(entityClass, attributes);
  return manager.save(entity);
}

type CreateAvatarAttributes = Partial<Avatar>;

export function createAvatar(attributes: CreateAvatarAttributes = {}) {
  return createEntity(Avatar, {
    imagePath: faker.random.arrayElement(AVATAR_IMAGES),
    color: faker.random.arrayElement(AVATAR_COLORS),
    ...attributes
  });
}

type CreateUserAttributes = Partial<User> & {
  password?: string;
  avatarAttributes?: CreateAvatarAttributes;
};

export async function createUser(attributes: CreateUserAttributes = {}) {
  const { avatarAttributes, password, ...userAttributes } = attributes;

  if (userAttributes.name === undefined) {
    userAttributes.name = faker.name.findName();
  }

  if (userAttributes.email === undefined) {
    userAttributes.email = faker.internet.email();
  }

  if (userAttributes.info === undefined) {
    userAttributes.info = faker.lorem.sentence();
  }

  if (userAttributes.avatarId === undefined) {
    const avatar = await createAvatar();
    userAttributes.avatarId = avatar.id;
  }

  if (avatarAttributes) {
    const avatar = await createAvatar(avatarAttributes);
    userAttributes.avatarId = avatar.id;
  }

  return createEntity(User, {
    passwordHash: hashPassword(password || "password"),
    isAdmin: false,
    ...userAttributes
  });
}

export function createAuthor(attributes: Partial<Author> = {}) {
  return createEntity(Author, {
    name: faker.name.findName(),
    bio: faker.lorem.sentence(),
    photoPath: faker.random.arrayElement(AUTHOR_PHOTOS),
    ...attributes
  });
}

type CreateBookAttributes = Partial<Book> & {
  authorAttributes?: Partial<Author>;
};

export async function createBook(attributes: CreateBookAttributes = {}) {
  const { authorAttributes, ...bookAttributes } = attributes;

  if (bookAttributes.authorId === undefined) {
    const author = await createAuthor(authorAttributes);
    bookAttributes.authorId = author.id;
  }

  return createEntity(Book, {
    title: titleizeSentence(
      faker.lorem.words(faker.random.number({ min: 1, max: 4 }))
    ),
    description: faker.lorem.sentence(),
    coverPath: faker.random.arrayElement(BOOK_COVERS),
    favourite: faker.random.boolean(),
    ...bookAttributes
  });
}

export async function createBookCopy(
  attributes: Partial<BookCopy> & {
    bookAttributes?: CreateBookAttributes;
  } & {
    ownerAttributes?: CreateUserAttributes;
  } & { borrowerAttributes?: CreateUserAttributes } = {}
) {
  const {
    bookAttributes,
    ownerAttributes,
    borrowerAttributes,
    ...bookCopyAttributes
  } = attributes;

  if (bookCopyAttributes.bookId === undefined) {
    const book = await createBook(bookAttributes);
    bookCopyAttributes.bookId = book.id;
  }

  if (bookCopyAttributes.ownerId === undefined) {
    const owner = await createUser(ownerAttributes);
    bookCopyAttributes.ownerId = owner.id;
  }

  if (borrowerAttributes) {
    const borrower = await createUser(borrowerAttributes);
    bookCopyAttributes.borrowerId = borrower.id;
  }

  return createEntity(BookCopy, bookCopyAttributes);
}
