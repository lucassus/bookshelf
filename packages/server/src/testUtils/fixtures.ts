import { getConnection } from "typeorm";

import { Author } from "../entity/Author";
import { Book } from "../entity/Book";
import { User } from "../entity/User";

async function loadAuthors() {
  const { manager } = getConnection();

  const authors = manager.create(Author, [
    {
      name: "J. K. Rowling",
      photoPath: "/images/book-authors/j-k-rowling.jpg"
    },
    {
      name: "James S. A. Corey",
      photoPath: "/images/book-authors/james-s-a-corey.jpg"
    },
    {
      name: "Andrzej Sapkowski",
      photoPath: "/images/book-authors/andrzej-sapkowski.jpg"
    }
  ]);

  return manager.save(authors);
}

function loadBooks(authors: Author[]) {
  const { manager } = getConnection();

  const books = manager.create(Book, [
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Sorcerer's Stone",
      coverPath: "/images/book-covers/harry1.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Chamber of Secrets",
      coverPath: "/images/book-covers/harry2.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Prisoner of Azkaban",
      coverPath: "/images/book-covers/harry3.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Goblet of Fire",
      coverPath: "/images/book-covers/harry4.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Order of the Phoenix",
      coverPath: "/images/book-covers/harry5.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Half-Blood Prince",
      coverPath: "/images/book-covers/harry6.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Deathly Hallows",
      coverPath: "/images/book-covers/harry7.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Leviathan Wakes",
      coverPath: "/images/book-covers/expanse1.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Caliban's War",
      coverPath: "/images/book-covers/expanse2.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Abaddon's Gate",
      coverPath: "/images/book-covers/expanse3.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Cibola Burn",
      coverPath: "/images/book-covers/expanse4.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Nemesis Games",
      coverPath: "/images/book-covers/expanse5.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Babylon's Ashes",
      coverPath: "/images/book-covers/expanse6.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Persepolis Rising",
      coverPath: "/images/book-covers/expanse7.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Tiamat's Wrath",
      coverPath: "/images/book-covers/expanse8.jpg"
    },
    {
      authorId: authors[2].id,
      title: "Blood of Elves",
      coverPath: "/images/book-covers/witcher1.jpg"
    },
    {
      authorId: authors[2].id,
      title: "Time of contempt",
      coverPath: "/images/book-covers/witcher2.jpg"
    },
    {
      authorId: authors[2].id,
      title: "Baptism of fire",
      coverPath: "/images/book-covers/witcher3.jpg"
    },
    {
      authorId: authors[2].id,
      title: "The tower of the swallow",
      coverPath: "/images/book-covers/witcher4.jpg"
    },
    {
      authorId: authors[2].id,
      title: "The lady of the lake",
      coverPath: "/images/book-covers/witcher5.jpg"
    }
  ]);

  return manager.save(books);
}

async function loadUsers() {
  const { manager } = getConnection();

  const users = manager.create(User, [
    {
      name: "Alice",
      email: "alice@example.com",
      avatar: {
        imagePath: "/images/avatars/w13.png",
        color: "yellow"
      }
    },
    {
      name: "Bob",
      email: "bob@example.com",
      avatar: {
        imagePath: "/images/avatars/m10.png",
        color: "green"
      }
    },
    {
      name: "Celine",
      email: "celine@example.com",
      avatar: {
        imagePath: "/images/avatars/w2.png",
        color: "red"
      }
    },
    {
      name: "Dan",
      email: "dan@example.com",
      avatar: {
        imagePath: "/images/avatars/m25.png",
        color: "blue"
      }
    }
  ]);
  await manager.save(users);
}

export const loadFixtures = async () => {
  const authors = await loadAuthors();
  await loadBooks(authors);

  await loadUsers();
};
