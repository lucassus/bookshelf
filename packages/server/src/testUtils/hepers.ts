import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";
import { getConnection } from "typeorm";

import { Context } from "../common/types";
import { ASSETS_BASE_URL } from "../config";
import { buildAuthorsLoader } from "../database/authorsLoader";
import { rootSchema } from "../graphql/rootSchema";

export const createTestClient = (
  context: Partial<Context> = {}
): ApolloServerTestClient => {
  const server = new ApolloServer({
    schema: rootSchema,
    context: {
      assetsBaseUrl: ASSETS_BASE_URL,
      authorsLoader: buildAuthorsLoader(),
      connection: getConnection(),
      ...context
    }
  });

  return createApolloTestClient(server);
};
