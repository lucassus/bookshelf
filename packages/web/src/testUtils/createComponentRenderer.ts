import { render } from "@testing-library/react";
import React from "react";

export function createComponentRenderer<CP>(
  Component: React.FunctionComponent<CP> | React.ComponentClass<CP>,
  baseProps: CP
) {
  return function renderComponent(props: Partial<CP> = {}) {
    const createElement = (newProps: Partial<CP> = {}) =>
      React.createElement(Component, {
        ...baseProps,
        ...props,
        ...newProps
      });

    const { rerender, ...rest } = render(createElement());

    return {
      rerender: (newProps: Partial<CP> = {}) =>
        rerender(createElement(newProps)),
      ...rest
    };
  };
}
