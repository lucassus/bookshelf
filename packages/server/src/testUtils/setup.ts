import { Connection } from "typeorm";

import { createDatabaseConnection } from "../createDatabaseConnection";
import { loadFixtures } from "./fixtures";

let connection: Connection;

beforeEach(async () => {
  connection = await createDatabaseConnection();
  await connection.dropDatabase();
  await connection.synchronize();
  await loadFixtures();
});

afterEach(() => connection?.close());
