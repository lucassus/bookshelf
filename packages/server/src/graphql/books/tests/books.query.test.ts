import { gql } from "apollo-server-express";

import { createTestClient } from "../../../testUtils/createTestClient";
import {
  createAuthor,
  createBook,
  createBookCopy,
  createUser
} from "../../../testUtils/factories";

describe("books query", () => {
  it("fetches books", async () => {
    // Given
    const book = await createBook({ title: "Hobbit" });
    await createBookCopy({ book, ownerAttributes: { name: "John" } });

    const borrower = await createUser({ name: "Paul" });
    await createBookCopy({ book, borrower });

    const author = await createAuthor({ name: "George Lucas" });
    await createBook({ author, title: "Star Wars IV" });
    await createBook({ author, title: "Star Wars V" });

    // When
    const res = await createTestClient().query({
      query: gql`
        query {
          booksCount
          books {
            id
            title
            cover {
              path
              url
            }
            copies {
              owner {
                id
                name
              }
              borrower {
                id
                name
              }
            }
          }
        }
      `
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).not.toBeNull();
    expect(res.data).toMatchSnapshot();
  });

  it("fetches books with authors", async () => {
    // Given
    const author = await createAuthor({ name: "Tolkien" });
    await createBook({ title: "Hobbit", author });
    await createBook({ title: "Lord of the Rings", author });
    await createBook();
    await createBook();

    // When
    const res = await createTestClient().query({
      query: gql`
        query {
          booksCount
          books {
            id
            title
            author {
              id
              name
            }
          }
        }
      `
    });

    // Then
    expect(res.data).not.toBeNull();
    expect(res.data).toMatchSnapshot();
  });

  it("fetches books along with authors and books again", async () => {
    // Given
    const author = await createAuthor({ name: "Andrzej Sapkowski" });
    await createBook({ author });
    await createBook({ author });

    // When
    const res = await createTestClient().query({
      query: gql`
        query {
          books {
            title
            author {
              name
              books {
                title
              }
            }
          }
        }
      `
    });

    // Then
    expect(res.data).not.toBeNull();
    expect(res.data).toMatchSnapshot();
  });
});
