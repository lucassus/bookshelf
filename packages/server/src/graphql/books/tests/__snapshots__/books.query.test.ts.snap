// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`books query fetches books 1`] = `
Object {
  "books": Array [
    Object {
      "copies": Array [
        Object {
          "borrower": null,
          "owner": Object {
            "id": "MS1Vc2Vy",
            "name": "John",
          },
        },
        Object {
          "borrower": Object {
            "id": "Mi1Vc2Vy",
            "name": "Paul",
          },
          "owner": Object {
            "id": "My1Vc2Vy",
            "name": "Gino Schuppe",
          },
        },
      ],
      "cover": Object {
        "path": "/images/book-covers/harry5.jpg",
        "url": "https://res.cloudinary.com/lucassus/image/upload/bookshelf/images/book-covers/harry5.jpg",
      },
      "id": "MS1Cb29r",
      "title": "Hobbit",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "path": "/images/book-covers/expanse1.jpg",
        "url": "https://res.cloudinary.com/lucassus/image/upload/bookshelf/images/book-covers/expanse1.jpg",
      },
      "id": "Mi1Cb29r",
      "title": "Star Wars IV",
    },
    Object {
      "copies": Array [],
      "cover": Object {
        "path": "/images/book-covers/expanse7.jpg",
        "url": "https://res.cloudinary.com/lucassus/image/upload/bookshelf/images/book-covers/expanse7.jpg",
      },
      "id": "My1Cb29r",
      "title": "Star Wars V",
    },
  ],
  "booksCount": 3,
}
`;

exports[`books query fetches books along with authors and books again 1`] = `
Object {
  "books": Array [
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Qui Adipisci Qui",
          },
          Object {
            "title": "Veniam",
          },
        ],
        "name": "Andrzej Sapkowski",
      },
      "title": "Qui Adipisci Qui",
    },
    Object {
      "author": Object {
        "books": Array [
          Object {
            "title": "Qui Adipisci Qui",
          },
          Object {
            "title": "Veniam",
          },
        ],
        "name": "Andrzej Sapkowski",
      },
      "title": "Veniam",
    },
  ],
}
`;

exports[`books query fetches books with authors 1`] = `
Object {
  "books": Array [
    Object {
      "author": Object {
        "id": "My1BdXRob3I=",
        "name": "Hayden Orn",
      },
      "id": "NC1Cb29r",
      "title": "Alias",
    },
    Object {
      "author": Object {
        "id": "Mi1BdXRob3I=",
        "name": "Velda Bailey",
      },
      "id": "My1Cb29r",
      "title": "Exercitationem Reiciendis Cupiditate Provident",
    },
    Object {
      "author": Object {
        "id": "MS1BdXRob3I=",
        "name": "Tolkien",
      },
      "id": "MS1Cb29r",
      "title": "Hobbit",
    },
    Object {
      "author": Object {
        "id": "MS1BdXRob3I=",
        "name": "Tolkien",
      },
      "id": "Mi1Cb29r",
      "title": "Lord of the Rings",
    },
  ],
  "booksCount": 4,
}
`;
